crates/red_knot/src/salsa_db/semantic/types.rs

```diff
@@ -0,0 +1,662 @@
+        .functions()
+        .map(|(id, _)| (id, FunctionTypingScope::new(db, file, id)))
+        .collect();
+    let classes = ast_ids
+        .classes()
+        .map(|(id, _)| (id, ClassTypingScope::new(db, file, id)))
+        .collect();
+
+    TypingScopes { functions, classes }
+}
+
+#[salsa::tracked(jar=Jar)]
+pub struct FunctionTypingScope {
+    file: File,
+    #[id]
+    id: FunctionId,
+}
+
+impl FunctionTypingScope {
+    pub fn node(self, db: &dyn Db) -> &StmtFunctionDef {
+        StmtFunctionDef::lookup(db, self.file(db), self.id(db))
+    }
+}
+
+#[salsa::tracked(jar=Jar)]
+pub struct ClassTypingScope {
+    file: File,
+    #[id]
+    id: ClassId,
+}
+
+impl ClassTypingScope {
+    pub fn node(self, db: &dyn Db) -> &StmtClassDef {
+        StmtClassDef::lookup(db, self.file(db), self.id(db))
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct TypingScopes {
+    functions: FxHashMap<FunctionId, FunctionTypingScope>,
+    classes: FxHashMap<ClassId, ClassTypingScope>,
+}
+
+impl std::ops::Index<FunctionId> for TypingScopes {
+    type Output = FunctionTypingScope;
+
+    fn index(&self, index: FunctionId) -> &Self::Output {
+        &self.functions[&index]
+    }
+}
+
+impl std::ops::Index<ClassId> for TypingScopes {
+    type Output = ClassTypingScope;
+
+    fn index(&self, index: ClassId) -> &Self::Output {
+        &self.classes[&index]
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum TypingScope {
+    Function(FunctionTypingScope),
+    Class(ClassTypingScope),
+    Module(File),
+}
+
+impl TypingScope {
+    pub fn for_symbol(db: &dyn Db, symbol: GlobalSymbolId) -> TypingScope {
+        let symbols = symbol_table(db, symbol.file());
+        let scope = symbols.scope_id_of_symbol(symbol.local());
+
+        Self::for_symbol_scope(db, GlobalId::new(symbol.file(), scope))
+    }
+
+    pub fn for_expression(db: &dyn Db, expression_id: GlobalId<ExpressionId>) -> TypingScope {
+        let symbols = symbol_table(db, expression_id.file());
+
+        let expression_scope = symbols.scope_id_of_expression(expression_id.local());
+        TypingScope::for_symbol_scope(db, GlobalId::new(expression_id.file(), expression_scope))
+    }
+
+    fn for_symbol_scope(db: &dyn Db, scope_id: GlobalId<ScopeId>) -> TypingScope {
+        let typing_scopes = typing_scopes(db, scope_id.file());
+        let symbols = symbol_table(db, scope_id.file());
+        let scope = symbols.scope(scope_id.local());
+
+        let mut ancestors = std::iter::once((scope_id.local(), scope))
+            .chain(symbols.parent_scopes(scope_id.local()));
+
+        let typing_scope = ancestors.find_map(|(_, scope)| match scope.definition()? {
+            Definition::Function(function) => Some(typing_scopes[*function].into()),
+            Definition::Class(class) => Some(typing_scopes[*class].into()),
+            _ => None,
+        });
+        typing_scope.unwrap_or(TypingScope::Module(scope_id.file()))
```

Comment:
What I understood is that we don't have a 1:1 mapping in case of lambda expressions and comprehensions. That's one reason why they're different.

The other reason is that salsa queries can only take Salsa ingredients (something stored in Salsa) as arguments, our `ScopeId`s aren't ingredient because they're created by pushing a value in an `IndexVec` which is more efficient than using a salsa interned struct which is always backed by a hash map.

I think it would make sense to use the same scope structure if we decide to build the symbol table per scope as well. Before then, I think it makes sense to keep them separate.